### PR TITLE
feat(cast): run arbitrary bytecode

### PR DESCRIPTION
**Pull request by abigger87**
## Overview

This PR finishes the work in #945  

## Motivation & Solution

This PR introduces a new subcommand for cast: `run` that executes arbitrary bytecode similar to `hevm exec`.

## Checklist

- [x] Parse Bytecode
- [x] Parse Calldata
- [x] Setup executor with bytecode
- [ ] Prettify raw call result
- [x] Unit Tests
- [x] CLI Docs

### CODE REVIEW

File: cli/src/cmd/cast/run.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        // Parse bytecode string
+        let bytecode_vec = self.bytecode.strip_prefix("0x").unwrap_or(&self.bytecode);
+        let parsed_bytecode = Bytes::from(hex::decode(bytecode_vec)?);
+
+        println!("Got bytecode: {:?}", parsed_bytecode.to_vec());
+
+        // Load figment
+        let figment: Figment = From::from(&self);
+        let evm_opts = figment.extract::<EvmOpts>()?;
+        let verbosity = 5; // evm_opts.verbosity;
```
Comment (gakonst) [inline cli/src/cmd/cast/run.rs:66]:
```suggestion
        let verbosity = evm_opts.verbosity;
```

Comment (abigger87) [reply cli/src/cmd/cast/run.rs:66]: Oops sorry meant to open as a draft!

File: cli/src/cmd/cast/run.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    fn run(self) -> eyre::Result<Self::Output> {
+        // Load figment
+        let figment: Figment = From::from(&self);
+        let evm_opts = figment.extract::<EvmOpts>()?;
+        let verbosity = evm_opts.verbosity;
+        let config = Config::from_provider(figment).sanitized();
+
+        // Parse bytecode string
+        let bytecode_vec = self.bytecode.strip_prefix("0x").unwrap_or(&self.bytecode);
+        let parsed_bytecode = Bytes::from(hex::decode(bytecode_vec)?);
```
Comment (mattsse) [inline cli/src/cmd/cast/run.rs:57]: I think this should work as well
```suggestion
        let bytecode_vec:Bytes = self.bytecode.parse()?;
```

File: cli/test-utils/src/macros.rs [hunk 3]
```diff
@@ -45,6 +45,20 @@ macro_rules! forgetest {
...
+        $crate::casttest!($test, $crate::ethers_solc::PathStyle::Dapptools, $fun);
+    };
+    ($test:ident, $style:expr, $fun:expr) => {
+        #[test]
+        fn $test() {
+            let (prj, cmd) = $crate::util::setup_cast(stringify!($test), $style);
+            $fun(prj, cmd);
+        }
+    };
+}
```
Comment (mattsse) [inline cli/test-utils/src/macros.rs:60]: nice!

File: cli/src/cmd/cast/run.rs (hunk 2 above)
Comment (abigger87) [reply cli/src/cmd/cast/run.rs]: done, thanks :)

**Comment by abigger87** [comment]
Right now, `cast run` deploys bytecode through the program setup and executes a raw call with the provided calldata - essentially it's `forge run` but against bytecode (rather than solidity contracts).

Open Questions:
Should `cast run` emulate `hevm exec`, that takes the bytecode as the program bytecode and executes against that (as shown below)?
Screenshot from 2022-03-22 16-49-42

Currently, there is a stark difference between the outputs of `cast run` and `hevm exec` as shown below:
Screenshot from 2022-03-22 16-51-30

**Comment by onbjerg** [comment]
I think this was auto-closed - rebase on top of master (which now includes revm changes) and re-open :smile:

**Comment by abigger87** [comment]
haha no worries!

**Comment by tynes** [comment]
Any plans to reopen this one?

**Pull request closed without merging**